### PR TITLE
UCT/TCP: Implement active messages.

### PR DESCRIPTION
**Pull request by yosefe**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5204/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5217/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5218/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5219/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5220/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5221/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5234/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5240/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5245/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5247/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5248/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@brminich @alex-mikheev pls take a look

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5250/ for details (Mellanox internal link).

File: src/uct/tcp/tcp_iface.c [hunk 1]
```diff
@@ -142,32 +181,31 @@ ucs_status_t uct_tcp_iface_set_sockopt(uct_tcp_iface_t *iface, int fd)
     return UCS_OK;
 }
 
+int64_t uct_tcp_addr_uuid(const struct sockaddr_in *in_addr)
```
Comment (brminich) [inline src/uct/tcp/tcp_iface.c]: will you add its usage in the next PR?

File: src/uct/tcp/tcp_ep.c [hunk 2]
```diff
@@ -50,11 +89,223 @@ static UCS_CLASS_INIT_FUNC(uct_tcp_ep_t, uct_iface_t *tl_iface,
...
+    uct_tcp_iface_t *iface = ucs_derived_of(ep->super.super.iface,
+                                            uct_tcp_iface_t);
+    uct_tcp_am_hdr_t *hdr;
+    ucs_status_t status;
+    size_t recv_length;
+    ssize_t remainder;
+
+    ucs_trace_func("ep=%p", ep);
+
+    /* Receive next chunk of data */
```
Comment (brminich) [inline src/uct/tcp/tcp_ep.c]: can it interfere with send?  Or i'm missing something

File: src/ucp/tag/rndv.c [hunk 3]
```diff
@@ -134,6 +134,8 @@ static size_t ucp_tag_rndv_rtr_pack(void *dest, void *arg)
     ucp_rndv_rtr_hdr_t *rndv_rtr_hdr = dest;
     size_t packed_len = sizeof(*rndv_rtr_hdr);
 
+    memset(rndv_rtr_hdr, 0, sizeof(*rndv_rtr_hdr));
```
[the remark this answers is not in the document]
Comment (yosefe) [reply src/ucp/tag/rndv.c]: yes, rtr packet was uninitialized and when using with TCP valgrind found it (it doesn't fail w/o tcp)

File: src/uct/tcp/tcp.h [hunk 4]
```diff
@@ -7,24 +7,37 @@
...
-KHASH_MAP_INIT_INT64(uct_tcp_fd_hash, uct_tcp_recv_sock_t*);
+/** How many events to wait for in epoll_wait */
+#define UCT_TCP_MAX_EVENTS        32
+
+
+/**
+ * TCP active message header
+ */
+typedef struct uct_tcp_am_hdr {
+    uint16_t                      am_id;
```
[the remark this answers is not in the document]
Comment (yosefe) [reply src/uct/tcp/tcp.h]: will reduce to 1

File: src/uct/tcp/tcp_ep.c (hunk 2 above)
Comment (yosefe) [reply src/uct/tcp/tcp_ep.c]: endpoint structure is used either for send or receive (but not both), so no

File: src/uct/tcp/tcp_iface.c (hunk 1 above)
Comment (yosefe) [reply src/uct/tcp/tcp_iface.c]: leftover, will remove

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5253/ for details (Mellanox internal link).

File: src/uct/tcp/tcp_ep.c [hunk 5]
```diff
@@ -50,11 +89,223 @@ static UCS_CLASS_INIT_FUNC(uct_tcp_ep_t, uct_iface_t *tl_iface,
...
+
+        uct_iface_trace_am(&iface->super, UCT_AM_TRACE_TYPE_RECV, hdr->am_id,
+                           hdr + 1, hdr->length, "RECV fd %d", ep->fd);
+        uct_iface_invoke_am(&iface->super, hdr->am_id, hdr + 1,
+                            hdr->length, 0);
+    }
+
+    /* Move the remaining data to the beginning of the buffer */
+    ucs_assert(remainder >= 0);
+    memmove(ep->buf, ep->buf + ep->offset, remainder);
```
Comment (alex-mikheev) [inline src/uct/tcp/tcp_ep.c]: TODO: remove extra copy on partial receive

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5255/ for details (Mellanox internal link).

**Pull request merged**